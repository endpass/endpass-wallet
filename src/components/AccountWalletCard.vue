<template>
  <div class="card app-card">
    <div class="card-header">
      <h2 class="card-header-title">Receive ETH</h2>
    </div>
    <div
      class="card-content"
      data-test="current-account"
    >
      <div
        v-if="isCurrentAccount"
        class="card-section"
      >
        Your Active Address:
      </div>
      <div class="card-section">
        <account
          v-if="address"
          :currency="activeCurrencyName"
          :address="address"
          :balance="balance"
        />
      </div>
      <div
        v-if="allowSend"
        class="card-section"
      >
        <v-button
          type="button"
          name="button"
          data-test="send-button"
          @click="emitSend"
        >
          Send ethereum
        </v-button>
      </div>
      <div class="card-tokens">
        <v-spinner v-if="isLoading" />
        <tokens-list
          v-if="!isLoading"
          :tokens="accountTokensList"
        />
      </div>
    </div>
  </div>
</template>

<script>
import { mapActions, mapGetters } from 'vuex';
import Account from '@/components/Account';
import TokensList from '@/components/TokensList';
import VSpinner from '@/components/ui/VSpinner';
import VButton from '@/components/ui/form/VButton';

export default {
  name: 'AccountWalletCard',

  props: {
    isCurrentAccount: {
      type: Boolean,
      default: false,
    },

    address: {
      type: String,
      required: true,
    },

    activeCurrencyName: {
      type: String,
      required: true,
    },

    balance: {
      type: String,
      default: '0',
    },

    allowSend: {
      type: Boolean,
      default: false,
    },
  },

  data: () => ({
    isLoading: false,
  }),

  computed: {
    ...mapGetters('tokens', ['fullTokensByAddress']),

    accountTokens() {
      return this.fullTokensByAddress(this.address);
    },

    accountTokensList() {
      return Object.values(this.accountTokens);
    },
  },

  methods: {
    ...mapActions('tokens', [
      'getTokensByAddress',
      'getTokensBalancesByAddress',
    ]),

    async loadTokensData() {
      const {
        address,
        accountTokens,
        getTokensByAddress,
        getTokensBalancesByAddress,
      } = this;

      this.isLoading = true;

      if (Object.keys(accountTokens).length === 0) {
        await getTokensByAddress({ address });
      }

      await getTokensBalancesByAddress({
        address,
      });

      this.isLoading = false;
    },

    emitSend() {
      this.$emit('send');
    },
  },

  created() {
    this.loadTokensData();
  },

  components: {
    Account,
    TokensList,
    VSpinner,
    VButton,
  },
};
</script>

<style lang="scss" scoped>
.card-section {
  margin-bottom: 1rem;
}

.card-tokens {
  position: relative;
}
</style>
