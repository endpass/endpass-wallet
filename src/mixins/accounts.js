// Mixin for common methods for Ethereum accounts
import web3 from 'web3';

const { fromWei } = web3.utils;

export default {
  computed: {
    address() {
      return this.$store.state.accounts.address;
    },

    balance() {
      return this.$store.state.accounts.balance === null
        ? null
        : fromWei(this.$store.getters['accounts/balance']);
    },
  },
};
