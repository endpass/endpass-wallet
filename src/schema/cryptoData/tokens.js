export const cryptoDataToken = {
  type: 'object',
  required: ['name', 'symbol', 'address'],
  properties: {
    address: {
      type: 'string',
      pattern: '^0x[0-9A-Fa-f]+$',
    },
    name: {
      type: 'string',
      minLength: 1,
    },
    symbol: {
      type: 'string',
      minLength: 1,
    },
    balance: {
      type: 'string',
      pattern: '[0-9]+',
      default: '0',
    },
    price: {
      type: 'boolean',
      default: false,
    },
  },
};

export const cryptoDataTokens = {
  type: 'array',
  items: cryptoDataToken,
};
