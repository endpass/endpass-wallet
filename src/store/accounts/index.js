import actions from './actions';
import getters from './getters';
import mutations from './mutations';

/**
 * @namespace
 * @property {Boolean} isInited Status of initializing accounts module
 * @property {Object} hdKey The encrypted xprv key for the user's hd wallet as a V3 keystore object.
 *  All accounts are direct child accounts.
 * @property {Object} hadrwareXpub The user's hardware xpub, keyed by wallet type
 * @property {Object} wallets The user's wallets, mapped by address
 * @property {String} address Current wallet address
 * @property {String} balance Currect account balance
 */
const state = {
  isInited: false,
  hardwareXpub: {},
  wallets: {},
  hdKey: null,
  address: null,
  balance: null,
};

export default {
  namespaced: true,
  state,
  actions,
  getters,
  mutations,
};
