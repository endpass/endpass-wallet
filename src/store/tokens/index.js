import actions from './actions';
import getters from './getters';
import mutations from './mutations';

export const state = {
  networkTokens: {},
  userTokens: {},
  prices: {},
  tokensByAddress: {},
  balancesByAddress: {},
  isLoading: false,
  intervalId: null,
};

export default {
  namespaced: true,
  state,
  actions,
  getters,
  mutations,
};
