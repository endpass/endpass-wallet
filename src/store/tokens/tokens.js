import TokenTracker from 'eth-token-tracker'
import EthplorerService from '@/services/ethplorer'

export default {
  namespaced: true,
  state() {
    let savedTokens = JSON.parse(localStorage.getItem('tokens') || '[]');
    return {
      // tokens from subscribtipn
      activeTokens: [],
      //tokens from localStorage
      savedTokens,
      tokensSubscription: null,
      tokensSerializeInterval: null
    }
  },
  mutations: {
    saveTokenToWatchStorage(state, token) {
      state.savedTokens.push(token);
      localStorage.setItem('tokens', JSON.stringify(state.savedTokens));
    },
    saveTokens(state, tokens) {
      // TODO check for errors here, activeTokens is undefined
      state.activeTokens = tokens || []
    },
    saveInterval(state, interval) {
      state.tokensSerializeInterval = interval;
    },
    saveSubscription(state, subscription) {
      state.tokensSubscription = subscription;
    }
  },
  actions: {
    addTokenToSubscribtion(context, token) {
      // Save token without blance for furer seances
      let tokenExist = context.state.tokensSubscription.tokens.find((subscribtionToken) => {
        return subscribtionToken.address === token.address;
      })
      if(!tokenExist) {
        context.commit('saveTokenToWatchStorage', token);
        context.state.tokensSubscription.add({
          address: token.address
        });
      }
    },
    subscribeOnTokenUpdates(context) {
      //destroy old subscription and recreate new one (in case of addres/provider change)
      if(context.rootState.accounts.activeAccount) {
        if(context.state.tokensSerializeInterval) {
          clearInterval(context.state.tokensSerializeInterval);
          context.state.tokensSubscription.stop();
        }
        // get tokens with balances
        context.dispatch('getNonZeroTokens').then((resp)=> {
          context.dispatch('createTokenSubscribtion', resp.data.tokens || []);
        });
      }
    },
    createTokenSubscribtion(context, nonZerotokens) {
      const address = context.rootState.accounts.activeAccount.getAddressString();
      //remove repetitive tokens
      const filteredSavedTokensTokens = context.state.savedTokens.filter((savedToken) => {
        return !nonZerotokens.find((nonZeroToken) => {
          return nonZeroToken.tokenInfo.address === savedToken.address;
        });
      });
      const tokensToWatch = filteredSavedTokensTokens.concat(nonZerotokens.map((nonZeroToken) => {
        return {
          address: nonZeroToken.tokenInfo.address
        }
      }));
      const subscription = new TokenTracker({
        userAddress: address,
        provider: context.rootState.web3.web3.currentProvider,
        pollingInterval: 4000,
        tokens: tokensToWatch
      });
      const interval = setInterval(()=> {
        if (context.state.tokensSubscription) {
          let balances = context.state.tokensSubscription.serialize();
          // TODO check for errors here
          if (balances.length && typeof balances[0].symbol !== 'undefined')
            context.commit('saveTokens', balances);
        }
      }, 4000);
      context.commit('saveInterval', interval);
      context.commit('saveSubscription', subscription);
    },
    getNonZeroTokens(context) {
      let address = context.rootState.accounts.activeAccount.getAddressString();
      return EthplorerService.getTransactions(address);
    }
  }
}
