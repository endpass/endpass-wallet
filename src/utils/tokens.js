import { toLower } from 'lodash';

export const makeConsistentToken = token => ({
  ...token,
  address: toLower(token.address),
});

export default {
  makeConsistentToken,
};
