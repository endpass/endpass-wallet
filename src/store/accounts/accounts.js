import { userService } from '@/services';
import storage from '@/services/storage';
import web3 from 'web3';
import Bip39 from 'bip39';
import HDKey from 'ethereumjs-wallet/hdkey';
import { hdKeyMnemonic, kdfParams } from '@/config';
import EthWallet from 'ethereumjs-wallet';
import { SAVE_TOKENS } from '@/store/tokens/mutations-types';
import { Wallet, Address } from '@/class';
import { BigNumber } from 'bignumber.js';
import keystore from '@/utils/keystore';

export default {
  namespaced: true,
  state: {
    // TODO move to user store?
    email: null,
    // The encrypted xprv key for the user's hd wallet as a V3 keystore
    // object. All accounts are direct child accounts.
    hdKey: null,
    // The user's wallets, keyed by address
    wallets: {},
    // The currently selected wallet
    wallet: null,
    address: null,
    balance: null,
    price: null,
    // prettier-ignore
    availableCurrencies: ['USD', 'AUD', 'BRL', 'CAD', 'CHF', 'CLP', 'CNY', 'CZK', 'DKK', 'EUR', 'GBP', 'HKD', 'HUF', 'IDR', 'ILS', 'INR', 'JPY', 'KRW', 'MXN', 'MYR', 'NOK', 'NZD', 'PHP', 'PKR', 'PLN', 'RUB', 'SEK', 'SGD', 'THB', 'TRY', 'TWD', 'ZAR'],
    settings: {
      fiatCurrency: 'USD',
    },
    otpSettings: {
      secret: null,
      status: null,
    },
  },
  getters: {
    isPublicAccount(state) {
      return (
        state.address instanceof Address && !(state.wallet instanceof Wallet)
      );
    },
    // TODO move to user store?
    isLoggedIn(state) {
      return !!state.email;
    },
    balance(state, getters, rootState, rootGetters) {
      if (state.balance === null) return null;

      const pendingBalance = rootGetters['transactions/pendingBalance'];
      const balanceWei = BigNumber(state.balance || '0')
        .minus(pendingBalance)
        .toFixed();

      return web3.utils.fromWei(balanceWei);
    },
    // Returns a decrypted HD Wallet
    hdWallet: state => password => {
      if (!state.hdKey) {
        return null;
      }
      return keystore.decryptHDWallet(password, state.hdKey);
    },
  },
  mutations: {
    setAddress(state, addressString) {
      state.address = new Address(addressString);
    },
    setWallet(state, wallet) {
      state.wallet = wallet;
    },
    addWallet(state, walletV3) {
      const wallet = new Wallet(walletV3);
      const { address } = walletV3;
      state.wallets = {
        ...state.wallets,
        [address]: wallet,
      };
    },
    // Saves the encrypted HD wallet key in V3 keystore format
    // Formerly addHdWallet
    setHdKey(state, key) {
      state.hdKey = key;
    },
    setBalance(state, balance) {
      state.balance = balance;
    },
    setPrice(state, price) {
      state.price = price;
    },
    setSettings(state, settings) {
      state.settings = JSON.parse(JSON.stringify(settings));
    },
    setEmail(state, email) {
      state.email = email;
    },
    setOtpSettings(state, otpSettings) {
      state.otpSettings = otpSettings;
    },
  },
  actions: {
    selectWallet({ commit, state, dispatch }, address) {
      commit('setWallet', state.wallets[address]);
      commit('setAddress', address);
      return dispatch(
        'tokens/subscribeOnTokensBalancesUpdates',
        {},
        { root: true },
      );
    },
    addWallet({ commit, dispatch, state }, json) {
      json.address = web3.utils.toChecksumAddress(json.address);
      commit('addWallet', json);

      return userService
        .setAccount(json.address, json)
        .catch(e => dispatch('errors/emitError', e, { root: true }));
    },
    addWalletAndSelect({ dispatch }, json) {
      return dispatch('addWallet', json)
        .then(() =>
          dispatch('selectWallet', web3.utils.toChecksumAddress(json.address)),
        )
        .catch(e => dispatch('errors/emitError', e, { root: true }));
    },
    // Import wallet from json V3 keystore
    async addWalletWithV3({ commit, dispatch }, { json, password }) {
      try {
        const wallet = new Wallet(json);
        let privateKey = await wallet.getPrivateKey(password);
        return dispatch('addWalletWithPrivateKey', privateKey);
      } catch (e) {
        return dispatch('errors/emitError', e, { root: true });
      }
    },
    async addWalletWithPrivateKey(
      { commit, dispatch },
      { privateKey, password },
    ) {
      try {
        const wallet = EthWallet.fromPrivateKey(Buffer.from(privateKey, 'hex'));
        const json = keystore.encryptWallet(password, wallet);
        return dispatch('addWalletAndSelect', json);
      } catch (e) {
        return dispatch('errors/emitError', e, { root: true });
      }
    },
    generateWallet({ commit, dispatch, state, getters }, password) {
      if (!state.hdKey) {
        return;
      }
      try {
        let hdWallet = getters.hdWallet(password);
        let i = Object.keys(state.wallets).length;
        let wallet = hdWallet.deriveChild(i).getWallet();
        let json = keystore.encryptWallet(password, wallet);
        return dispatch('addWalletAndSelect', json);
      } catch (e) {
        return dispatch('errors/emitError', e, { root: true });
      }
    },
    // Saves HD wallet's extended keys on the server
    saveHdWallet({ commit, dispatch, state }, json) {
      return userService.setAccount(json.address, json);
    },
    addHdWallet({ commit, dispatch }, { key, password }) {
      const seed = Bip39.mnemonicToSeed(key);
      const hdKey = HDKey.fromMasterSeed(seed);
      const hdWallet = hdKey.derivePath(hdKeyMnemonic.path);
      // Encrypt extended private key
      const json = keystore.encryptHDWallet(password, hdWallet);
      commit('setHdKey', json);

      // Save HD keys and generate the first child wallet
      return dispatch('saveHdWallet', json)
        .then(() => dispatch('generateWallet', password))
        .catch(e => dispatch('errors/emitError', e, { root: true }));
    },
    async addMultiHdWallet({ commit, dispatch, rootState }, { key, password }) {
      const seed = Bip39.mnemonicToSeed(key);
      const hdKey = HDKey.fromMasterSeed(seed);
      const hdWallet = hdKey.derivePath(hdKeyMnemonic.path);

      /* eslint-disable no-await-in-loop */
      for (let index = 0; index < 5; index++) {
        const wallet = hdWallet.deriveChild(index).getWallet();
        const walletV3 = wallet.toV3(Buffer.from(password), kdfParams);
        const { address } = walletV3;

        if (index !== 0) {
          dispatch('addWallet', walletV3);
        } else {
          dispatch('addWalletAndSelect', walletV3);
        }

        try {
          const balance = await rootState.web3.web3.eth.getBalance(address);

          if (balance === '0') {
            break;
          }
        } catch (e) {
          break;
        }
      }
      /* eslint-enable no-await-in-loop */
    },
    updateBalance({ commit, dispatch, state, rootState }) {
      if (state.address) {
        const address = state.address.getChecksumAddressString();

        return rootState.web3.web3.eth
          .getBalance(address)
          .then(balance => commit('setBalance', balance))
          .catch(e => dispatch('errors/emitError', e, { root: true }));
      }
    },
    updateSettings({ commit, dispatch }, settings) {
      commit('setSettings', settings);
      return storage
        .write('settings', settings)
        .catch(e => dispatch('errors/emitError', e, { root: true }));
    },
    validatePassword({ state }, password) {
      return state.wallet.validatePassword(password);
    },
    login({ commit, dispatch }, email) {
      return userService.login(email);
    },
    logout({ commit, dispatch }) {
      commit('setEmail', null);
      return Promise.all([storage.clear(), userService.logout()])
        .then(() => window.location.reload())
        .catch(e => dispatch('errors/emitError', e, { root: true }));
    },
    loginViaOTP({}, { code, email }) {
      return userService.loginViaOTP(code, email);
    },
    getOtpSettings({ commit, dispatch }) {
      return userService
        .getOtpSettings()
        .then(otpSettings => commit('setOtpSettings', otpSettings))
        .catch(e => dispatch('errors/emitError', e, { root: true }));
    },
    setOtpSettings({ commit, dispatch }, { secret, code }) {
      return userService
        .setOtpSettings(secret, code)
        .then(() => commit('setOtpSettings', { status: 'enabled' }))
        .catch(e => dispatch('errors/emitError', e, { root: true }));
    },
    deleteOtpSettings({ commit, dispatch }, { code }) {
      return userService
        .deleteOtpSettings(code)
        .then(() => {
          commit('setOtpSettings', {});
          dispatch('getOtpSettings');
        })
        .catch(e => dispatch('errors/emitError', e, { root: true }));
    },
    async init({ commit, dispatch }) {
      try {
        let [settings, email] = await Promise.all([
          userService.getSettings(),
          storage.read('email'),
        ]);

        commit(`tokens/${SAVE_TOKENS}`, settings.tokens || {}, { root: true });
        commit('setEmail', email);

        if (settings) {
          commit('setSettings', settings);
        }

        if (!email) {
          storage.disableRemote();
          return null;
        }

        // Fetch and save HD wallet
        let hdKey = await userService.getHDKey();
        if (hdKey) {
          commit('setHdKey', hdKey);
        }

        // Fetch and save regular accounts
        let accounts = await userService.getV3Accounts();
        if (accounts && accounts.length) {
          accounts.forEach(wallet => commit('addWallet', wallet));
          await dispatch('selectWallet', accounts[0].address);
        }
      } catch (e) {
        return dispatch('errors/emitError', e, { root: true });
      }
    },
  },
};
