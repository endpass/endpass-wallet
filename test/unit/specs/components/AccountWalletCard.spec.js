import Vuex from 'vuex';
import { shallowMount, createLocalVue } from '@vue/test-utils';
import UIComponents from '@endpass/ui';

import AccountWalletCard from '@/components/AccountWalletCard';

import { address } from 'fixtures/accounts';
import { tokensMappedByAddresses } from 'fixtures/tokens';

const localVue = createLocalVue();
localVue.use(Vuex);
localVue.use(UIComponents);

describe('AccountWalletCard', () => {
  let wrapper;
  let store;
  let options;
  let tokensActions;

  beforeEach(() => {
    tokensActions = {
      getTokensByAddress: jest.fn(),
      getTokensBalancesByAddress: jest.fn(),
    };
    store = new Vuex.Store({
      modules: {
        tokens: {
          namespaced: true,
          getters: {
            fullTokensByAddress: jest.fn(() => () => ({})),
          },
          actions: tokensActions,
        },
      },
    });
    options = {
      propsData: {
        activeCurrencyName: 'ETH',
        balance: '20',
        address,
      },
      store,
      localVue,
    };

    jest.clearAllMocks();
  });

  describe('render', () => {
    beforeEach(() => {
      wrapper = shallowMount(AccountWalletCard, {
        ...options,
      });
    });

    it('should be a Vue component', () => {
      expect(wrapper.isVueInstance()).toBeTruthy();
      expect(wrapper.name()).toBe('AccountWalletCard');
    });

    it('should render initial state of the component', () => {
      expect(wrapper.html()).toMatchSnapshot();
    });

    it('should correctly render component with current account', async () => {
      expect.assertions(1);

      wrapper = shallowMount(AccountWalletCard, {
        ...options,
        computed: {
          accountTokensList: () => [],
        },
      });
      wrapper.setProps({
        isCurrentAccount: true,
      });

      await wrapper.vm.$nextTick();

      expect(wrapper.html()).toMatchSnapshot();
    });

    it('should correctly render component with send button', () => {
      wrapper.setProps({
        allowSend: true,
      });

      expect(wrapper.html()).toMatchSnapshot();
    });

    it('should not render tokens list during loading', () => {
      wrapper.setData({
        isLoading: true,
      });

      expect(wrapper.find('tokens-list-stub').exists()).toBe(false);
    });

    it('should render tokens list if it is not loading', () => {
      wrapper.setData({
        isLoading: false,
      });

      expect(wrapper.find('tokens-list-stub').exists()).toBe(true);
    });
  });

  describe('behavior', () => {
    it('should load tokens data on component create if tokens are empty', async () => {
      expect.assertions(2);

      wrapper = shallowMount(AccountWalletCard, {
        ...options,
        computed: {
          accountTokens: () => ({}),
        },
      });

      await global.flushPromises();

      expect(tokensActions.getTokensByAddress).toBeCalled();
      expect(tokensActions.getTokensBalancesByAddress).toBeCalled();
    });

    it('should not load tokens data on component create if tokens are not empty', async () => {
      expect.assertions(2);

      wrapper = shallowMount(AccountWalletCard, {
        ...options,
        computed: {
          accountTokens: () => tokensMappedByAddresses,
        },
      });

      await global.flushPromises();

      expect(tokensActions.getTokensByAddress).not.toBeCalled();
      expect(tokensActions.getTokensBalancesByAddress).toBeCalled();
    });
  });
});
