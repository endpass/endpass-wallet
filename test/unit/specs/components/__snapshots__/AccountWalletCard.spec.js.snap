// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AccountWalletCard render should correctly render component with current account 1`] = `
<div class="card app-card">
    <div class="card-header">
        <h2 class="card-header-title">Receive ETH</h2>
    </div>
    <div data-test="current-account" class="card-content">
        <div class="card-section">
            Your Active Address:
        </div>
        <div class="card-section">
            <account currency="ETH" address="0xB14Ab53E38DA1C172f877DBC6d65e4a1B0474C3c" balance="20"></account>
        </div>
        <!---->
        <div class="card-tokens">
            <!---->
            <tokens-list tokens=""></tokens-list>
        </div>
    </div>
</div>
`;

exports[`AccountWalletCard render should correctly render component with send button 1`] = `
<div class="card app-card">
    <div class="card-header">
        <h2 class="card-header-title">Receive ETH</h2>
    </div>
    <div data-test="current-account" class="card-content">
        <!---->
        <div class="card-section">
            <account currency="ETH" address="0xB14Ab53E38DA1C172f877DBC6d65e4a1B0474C3c" balance="20"></account>
        </div>
        <div class="card-section">
            <v-button type="button" name="button" data-test="send-button" v-on:click="bound emitSend">
                Send ethereum
            </v-button>
        </div>
        <div class="card-tokens">
            <!---->
            <tokens-list tokens=""></tokens-list>
        </div>
    </div>
</div>
`;

exports[`AccountWalletCard render should render initial state of the component 1`] = `
<div class="card app-card">
    <div class="card-header">
        <h2 class="card-header-title">Receive ETH</h2>
    </div>
    <div data-test="current-account" class="card-content">
        <!---->
        <div class="card-section">
            <account currency="ETH" address="0xB14Ab53E38DA1C172f877DBC6d65e4a1B0474C3c" balance="20"></account>
        </div>
        <!---->
        <div class="card-tokens">
            <!---->
            <tokens-list tokens=""></tokens-list>
        </div>
    </div>
</div>
`;
